### DISCUSSION TITLE
Brew Counter

### DISCUSSION THREAD

**Discussion by aneisch** [TillFleisch/ESPHome-Philips-Smart-Coffee#76] [Show and tell]
Hacked together a brew counter for maintenance tracking purposes that others may find helpful:

```
text_sensor:
  - platform: philips_coffee_machine
    controller_id: philip
    id: status
    name: "${friendly_name} Status"
    on_value:
      then:
        - lambda: |-
            if (id(status).state.find("Brewing") != std::string::npos)  {
              auto call = id(brew_count).make_call();
              call.number_increment(true);
              call.perform();
            } 

number:
  - platform: template
    id: brew_count
    name: "${friendly_name} Brew Count"
    step: 1
    restore_value: True
    min_value: 0
    max_value: 99999
    optimistic: true
```

**Comment by quack3d** [comment]
Neat. I was planning on doing it in HA instead but there were a couple minor issues. For instance I didn't want to count every Brewing* since it will count twice if water is empty mid brew and you continue after you've filled the tank. So maybe it would work better to check when state goes from *selected to Brewing*. But then if you're quick with the presses it won't register what you've selected. So in those cases it will go from Idle -> Brewing* and not register the count. Maybe it's better to count when it goes from Brewing* -> [everything else than Water empty]?

**Reply by TillFleisch** [reply]
> For instance I didn't want to count every Brewing* since it will count twice if water is empty mid brew and you continue after you've filled the tank.

First of all, it's ridiculous that the machine does this. It's guesstimation work regarding the water level should be more conservative.
A better approach would be a lockout time, where the counter is only incremented if it has not been in the past `x` seconds/minutes. A good lockout time would be the brew-time of the quickest beverage.

**Reply by quack3d** [reply]
> A better approach would be a lockout time, where the counter is only incremented if it has not been in the past `x` seconds/minutes. A good lockout time would be the brew-time of the quickest beverage.

Maybe I'm not following entirely, but wouldn't this be an issue with that: Say you start brewing and it stops because of water empty. Then it will allow you up to maybe a minute (maybe even more if you take out the tank and THEN wait?) to refill tank and resume brewing. I don't know but I'm guessing a small espresso takes shorter than a minute. So if Espresso takes 40 seconds and you use that as lockout time, waiting 50 seconds before resuming after water empty will count twice.

Discussing edge cases is fun. 🤣

**Comment by quack3d** [comment]
Btw, wouldn't it be better to have it as a sensor rather than number? Now it's possible to change it in the UI since it's a number. Also, can you modify it so it counts twice if you select a 2x beverage when status is "Brewing 2x*"? 😊

**Comment by quack3d** [comment]
This seems to work nicely (but will register 2 brewings if water becomes empty while brewing and then resuming). I added one sensor for total brew count and another for service interval. The plan is to reset the service interval when reaching 500 (using the number component). Globals were needed to be able to use number components to manually update the sensors. And I use sensors in addition to numbers because they should work with long term statistics in HA.

```
globals:
  - id: service_interval_counter
    type: int
    restore_value: true
    #initial_value: '0'
  - id: brew_count
    type: int
    restore_value: true
    #initial_value: '0'  

text_sensor:
  - platform: philips_coffee_machine
    controller_id: philips
    id: status
    name: "Status"
    on_value:
      then:
        - lambda: |-
            if (id(status).state.find("Brewing 2x") != std::string::npos) {
              id(brew_count) += 2;
              id(service_interval_counter) += 2;
            } else if (id(status).state.find("Brewing") != std::string::npos) {
              id(brew_count) += 1;
              id(service_interval_counter) += 1;
            }
            auto brew_call = id(brew_count_number).make_call();
            brew_call.set_value(id(brew_count));
            brew_call.perform();

            auto service_call = id(service_interval_counter_number).make_call();
            service_call.set_value(id(service_interval_counter));
            service_call.perform();  

number:
  - platform: template
    id: service_interval_counter_number
    name: "Adjust Service Interval Counter"
    step: 1
    restore_value: true
    min_value: 0
    max_value: 99999
    optimistic: true
    set_action:
      then:
        - lambda: |-
            id(service_interval_counter) = (int) x;

  - platform: template
    id: brew_count_number
    name: "Adjust Brew Count"
    step: 1
    restore_value: true
    min_value: 0
    max_value: 99999
    optimistic: true
    set_action:
      then:
        - lambda: |-
            id(brew_count) = (int) x;

sensor:
  - platform: template
    name: "Service Interval Counter"
    lambda: |-
      return id(service_interval_counter);
    state_class: measurement
    unit_of_measurement: "count"
    accuracy_decimals: 0

  - platform: template
    name: "Brew Count"
    lambda: |-
      return id(brew_count);
    state_class: measurement
    unit_of_measurement: "count"
    accuracy_decimals: 0            			
			
```

**Reply by TillFleisch** [reply]
A template button or service (see [docs](https://esphome.io/components/api.html)) would probably be better suited for resetting the counter and it would remove the need for number components.

**Reply by quack3d** [reply]
You're right. I changed it to: 
```
globals:
  - id: service_interval_counter
    type: int
    restore_value: true
    #initial_value: '0'
  - id: brew_count
    type: int
    restore_value: true
    #initial_value: '0'  

text_sensor:
  - platform: philips_coffee_machine
    controller_id: philips
    id: status
    name: "Status"
    on_value:
      then:
        - lambda: |-
            if (id(status).state.find("Brewing 2x") != std::string::npos) {
              id(brew_count) += 2;
              id(service_interval_counter) += 2;
            } else if (id(status).state.find("Brewing") != std::string::npos) {
              id(brew_count) += 1;
              id(service_interval_counter) += 1;
            }       
			
button:
  - platform: template
    name: "Reset Service Interval Counter"
    on_press:
      then:
        - lambda: |-
            id(service_interval_counter) = 0;

  - platform: template
    name: "Reset Brew Count"
    on_press:
      then:
        - lambda: |-
            id(brew_count) = 0;			
			
sensor:
  - platform: template
    name: "Service Interval Counter"
    lambda: |-
      return id(service_interval_counter);
    state_class: measurement
    unit_of_measurement: "count"
    accuracy_decimals: 0

  - platform: template
    name: "Brew Count"
    lambda: |-
      return id(brew_count);
    state_class: measurement
    unit_of_measurement: "count"
    accuracy_decimals: 0     			
```
